splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/TriggerHandler.java

```diff
@@ -62,36 +75,174 @@
+                          ExecRow templateRow,
+                          String tableVersion) throws StandardException {
         WriteCursorConstantOperation constantAction = (WriteCursorConstantOperation) writeInfo.getConstantAction();
         initConnectionContext(activation.getLanguageConnectionContext());
 
         this.beforeEvent = beforeEvent;
         this.afterEvent = afterEvent;
+        this.activation = activation;
         this.resultDescription = activation.getResultDescription();
         this.pendingAfterRows = Lists.newArrayListWithCapacity(AFTER_ROW_BUFFER_SIZE);
 
         this.hasBeforeRow = triggerInfo.hasBeforeRowTrigger();
         this.hasAfterRow = triggerInfo.hasAfterRowTrigger();
         this.hasBeforeStatement = triggerInfo.hasBeforeStatementTrigger();
         this.hasAfterStatement = triggerInfo.hasAfterStatementTrigger();
+        this.hasStatementTriggerWithReferencingClause = triggerInfo.hasStatementTriggerWithReferencingClause();
         this.heapList = heapList;
+        this.templateRow = templateRow;
+        this.tableVersion = tableVersion;
+        this.activation = activation;
+        this.writeInfo = writeInfo;
         initTriggerActivator(activation, constantAction);
     }
 
+    public Callable<Void> getTriggerTempTableflushCallback () {
+        if (triggerRowHolder != null)
+             return triggerRowHolder.getTriggerTempTableflushCallback();
+        return null;
+    }
+
+    public void setTxn(TxnView txn) {
+        this.txn = txn;
+        if (triggerRowHolder != null)
+            triggerRowHolder.setTxn(txn);
+    }
+
+    public boolean hasStatementTrigger() {
+        return hasBeforeStatement || hasAfterStatement;
+    }
+
+    public boolean hasStatementTriggerWithReferencingClause() {
+        return this.hasStatementTriggerWithReferencingClause;
+    }
+
+    public void addRowToNewTableRowHolder(ExecRow row, KVPair encode) throws StandardException {
+        if (triggerRowHolder != null)
+            triggerRowHolder.insert(row, encode);
+    }
+
+    public void setIsSpark(boolean isSpark) {
+        this.isSpark = isSpark;
+    }
+
+    public boolean isSpark() {
+        return this.isSpark;
+    }
+
+    public long getTriggerConglomID() {
+        if (triggerRowHolder != null) {
+            return triggerRowHolder.getConglomerateId();
+        }
+        return 0;
+    }
+
+    public void initTriggerRowHolders(boolean isSpark, TxnView txn, byte[] token, long ConglomID) throws StandardException {
+        this.isSpark = isSpark;
+        this.txn = txn;
+        this.token = token;
+        Properties properties = new Properties();
+        if (hasStatementTriggerWithReferencingClause) {
+            // Use the smaller of ControlExecutionRowLimit or 1000000 to determine when to switch to spark execution.
+            // Hard cap at 1 million despite the setting of controlExecutionRowLimit since we don't want to exhaust
```

Comment:
Should that 1 million limit also be applied in the other place we use that setting?
If so, you can set the limit directly in SQLConfiguration.java
cf SPARK_COMPACTION_RESOLUTION_SHARE in OlapConfigurations.java for an example

Reply:
No, I would like users to be able to freely adjust that setting for normal SQL queries.  For triggers we have a hard upper limit of 1000000 rows for the in-memory buffer, which I don't want to grow any larger.  Instead we just switch to running on spark.